scripts/bigquery.py

```diff
@@ -0,0 +1,392 @@
+
+
+def transpile_schemas(output_path: Path, schema_paths: List[Path]):
+    """Write schemas to directory."""
+    assert output_path.is_dir()
+    for path in schema_paths:
+        namespace, doctype, filename = path.parts[-3:]
+        version = int(filename.split(".")[-3])
+        # pioneer-study schemas were done incorrectly and are ignored here
+        if namespace == "schemas":
+            print(f"skipping {path} due to wrong directory level")
+            continue
+        out = output_path / f"{namespace}.{doctype}.{version}.bq"
+        with out.open("w") as fp:
+            print(f"writing {out}")
+            json.dump(transpile(path), fp, indent=2)
+            fp.write("\n")
+
+
+def load_schemas(input_path: Path):
+    """Load schemas into memory for use in google-cloud-bigquery."""
+    paths = list(input_path.glob("*.bq"))
+    assert len(paths) > 0
+    schemas = {}
+    for path in paths:
+        qualified_name = path.parts[-1][:-3]
+        with path.open("r") as fp:
+            schemas[qualified_name] = json.load(fp)
+    print(f"loaded {len(schemas.keys())} schemas")
+    return schemas
+
+
+def git_stash_size():
+    """Find the size of the git stash."""
+    return len([item for item in run("git stash list").split("\n") if item])
+
+
+def resolve_ref(ref: str) -> str:
+    """Return a resolved reference or the short revision if empty."""
+    resolved = run(f"git rev-parse --abbrev-ref {ref}") or run(
+        f"git rev-parse --short {ref}"
+    )
+    if resolved != ref:
+        print(f"resolved {ref} to {resolved}")
+    return resolved
+
+
+@contextmanager
+def managed_git_state():
+    """Save the current git state.
+
+    Stash any changes so we can reference by real changes in the tree. If the
+    branch has in-flight changes, the changes would be ignored by the stash.
+    """
+    original_ref = run("git rev-parse --abbrev-ref HEAD")
+    before_stash_size = git_stash_size()
+    run("git stash")
+    should_apply_stash = before_stash_size != git_stash_size()
+    if should_apply_stash:
+        print(
+            "NOTE: uncommitted have been detected. These will be ignored during comparisons."
+        )
+    try:
+        yield
+    finally:
+        run(f"git checkout {original_ref}")
+        if should_apply_stash:
+            run("git stash apply")
+            # if apply fails, then an exception is thrown and the stash still
+            # exists
+            run("git stash drop")
+
+
+def _checkout_transpile_schemas(schemas: Path, ref: str, output: Path) -> Path:
+    """Checkout a revision, transpile schemas, and return to the original revision.
+
+    Generates a new folder under output with the short revision of the reference.
+    """
+    # preconditions
+    assert output.is_dir(), f"output must be a directory: {output}"
+    assert (
+        len(run("git diff")) == 0
+    ), f"current git state must be clean, please stash changes"
+
+    rev = run(f"git rev-parse --short {ref}")
```

Comment:
It could, but the revision is less ambiguous for storing the results.  For example, running this on `HEAD`, committing new changes, and running this again would result in a folder with the resolved branch name.